src/net/http_network.rs

```diff
@@ -18,74 +24,78 @@ use tokio_stream::wrappers::ReceiverStream;
 pub struct HttpNetwork {
     query_id: QueryId,
     sink_sender: mpsc::Sender<MessageChunks>,
-    sink_receiver: Arc<Mutex<Option<mpsc::Receiver<MessageChunks>>>>,
     message_stream_sender: mpsc::Sender<MessageChunks>,
     message_stream_receiver: Arc<Mutex<Option<mpsc::Receiver<MessageChunks>>>>,
 }
 
 impl HttpNetwork {
+    /// * Creates an [`HttpNetwork`]
+    /// * spawns a task that consumes incoming data from the infra layer intended for other helpers
     #[must_use]
     #[allow(unused)]
-    pub fn new<D: PeerDiscovery>(peer_discovery: &D, query_id: QueryId) -> Self {
-        let (stx, srx) = mpsc::channel(1);
+    pub fn new<D: PeerDiscovery>(role: Role, peer_discovery: &D, query_id: QueryId) -> Self {
+        let (stx, mut srx) = mpsc::channel::<MessageChunks>(1);
         let (mstx, msrx) = mpsc::channel(1);
-        let network = HttpNetwork {
+
+        let peers_conf = peer_discovery.peers();
+        let clients = Self::clients(role, peers_conf);
+        tokio::spawn(async move {
+            let mut last_seen_messages = HashMap::new();
+
+            while let Some((channel_id, messages)) = srx.recv().await {
+                // increment `offset` each time, per `ChannelId`
+                let offset = last_seen_messages.entry(channel_id.clone()).or_default();
+                let args = HttpSendMessagesArgs {
+                    query_id,
+                    step: &channel_id.step,
+                    offset: *offset,
+                    messages: Bytes::from(messages),
+                };
+                *offset += 1;
+
+                clients[channel_id.role]
+                    .send_messages(args)
+                    .await
+                    .unwrap_or_else(|err| {
+                        tracing::error!("could not send message to client: {err}");
```

Comment:
do you really want to swallow error here? If you fail to send, it is very unlikely that helpers can continue computation, but for some time system will pretend that everything is fine. If you need to build resiliency here, you could try to plug in backchannel to every message that you need to send (oneshot channel works here) and signal through it that something went wrong on the send side. Then, you'll leave it up to the sender to decide what they want to do - abort or retry.

Having said that, I don't think we need it right now and I would suggest to just panic here

Reply:
yeah, you're right. i will just panic.